### PR TITLE
[https://nvbugs/6460072][fix] Split the caller — for `use_lm_head_tp_in_adp and is_all_greedy_sample`, keep…

### PR DESCRIPTION
**Pull request by trtllm-agent**
## Summary
- **Root cause**: MTP-Eagle greedy path under ADP+LM-head-TP was trimming vocab-sharded logits to token_count and then plain-argmaxing per rank, so every rank picked a local-vocab-shard token (not the global argmax) and rows [:token_count] weren't even this rank's real requests (allgather-along-dim=0 order).
- **Fix**: Split the caller — for `use_lm_head_tp_in_adp and is_all_greedy_sample`, keep padded logits, call `self.draft_sampler(logits, draft_model.mtp_layers[0].shared_head.mapping_lm_head_tp)` (which all-gathers local argmax/max across mapping_lm_head_tp ranks and picks global argmax), then trim result to token_count. Non-greedy path unchanged.
- Automated fix generated by repair-bot

## Test plan
- [x] Verify fix on the same GPU type as the original failure
- [x] Check for regressions in related tests

## Links
- Bug: https://nvbugs/6460072

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved greedy token selection for speculative decoding when tensor-parallel language-head processing is enabled.
  * Ensured draft tokens are sampled from correctly sharded logits, improving sampling correctness and consistency.
* **Refactor**
  * Clarified internal handling and documentation for greedy sampling across parallel decoding modes.

### CODE REVIEW

File: tensorrt_llm/_torch/speculative/eagle3.py [hunk 1]
```diff
@@ -896,22 +896,26 @@ def _forward_linear_draft_loop(self, inputs, attn_metadata, spec_metadata,
...
-                                                     batch_size,
-                                                     draft_step=i)
+                # ADP+LM-head-TP: logits are vocab-sharded and batch dim is
+                # max_num_requests*lm_head_tp_size (allgather order), so a plain
+                # per-rank argmax on rows[:token_count] would pick the wrong
+                # rank's rows and only see a local vocab shard. Greedy: route
+                # through draft_sampler with mapping_lm_head_tp (global argmax
+                # via allgathered (idx,max)) then trim. Non-greedy: trim first
+                # so per-request sampling params (sized to token_count) align.
+                if use_lm_head_tp_in_adp and spec_metadata.is_all_greedy_sample:
```
Comment (xwang233) [inline tensorrt_llm/_torch/speculative/eagle3.py]: although not currently existing in main, there is a latent risk: if some ranks execute `draft_sampler` and other ranks execute `draft_decoder`, the all-gather calls within could hang

Comment (zhaoyangwang-nvidia) [reply tensorrt_llm/_torch/speculative/eagle3.py]: Good catch — that rank-divergence hazard is real for the earlier shape of this fix, where the all-gather was gated on spec_metadata.is_all_greedy_sample, a per-rank batch property. Under attention DP each rank holds different requests, so the flag can differ within an LM-head-TP group and the group all-gather would lose participants and hang.

The updated fix avoids this by construction: the collective is now gated only on mapping_lm_head_tp is not None && tp_size > 1, which is pure config (identical and constant across ranks). Since rejection sampling is config-gated off under attention DP, advanced is always False there, so every rank in the group unconditionally takes the same greedy_sample_draft_with_tp_gather all-gather branch on every draft step, regardless of its batch's sampling params. The payload shape is also rank-invariant (lm_head_tp_size * max_num_requests rows, 2 scalars per row), so both attendance and shapes always match.

**Review by zhaoyangwang-nvidia — APPROVED** [review]
